src/julia_gc.c

```diff
@@ -387,6 +388,36 @@ void MarkJuliaObj(void * obj)
 // from which freed objects are then marked. Hence, we add additional checks
 // when traversing GAP master pointer and bag objects that this happens
 // only for live objects.
+//
+// We use "bottom" to refer to the origin of the stack, and "top" to describe
+// the current stack pointer. Confusingly, on most contemporary architectures,
+// the stack grows "downwards", which means that the "bottom" of the stack is
+// the highest address and "top" is the lowest. When we treat the stack as an
+// abstract buffer, "start" and "end" refer to its lowest and highest address,
+// respectively.
+//
+//   +------------------------------------------------+
+//   | guard |    unused area          | active stack |
+//   | pages |  <--- growth ---        | frames       |
+//   +------------------------------------------------+
+//   ^                                 ^              ^
+//   |                                 |              |
+// start                              top         bottom/end
+//
+// All stacks in Julia are associated with tasks and we can use
+// jl_task_stack_buffer() to retrieve the buffer information for that stack.
+// That said, in a couple of cases these need adjustments.
+//
+// 1. The stack buffer of the root task of the main thread, when started
+//    from GAP can extend past the point where Julia believes its bottom is.
+//    Therefore, for that stack, we use GapBottomStack instead.
+// 2. If it is the current task of the current thread, we know where exactly
+//    the top is and do not need to scan the entire stack buffer.
+//
+// In addition, stack buffers can include guard pages, which trigger a
+// segmentation fault when accessed. We intercept such segmentation faults
+// and terminate stack scanning early for those.
```

Comment:
From this, it does not become clear (a) where the guard pages are (your sketch above of course shows this, but perhaps mention it again), but more importantly (b) it is unclear why Julia would point <start> beyond the guard page... And it's not clear to me if it really *does* that? I thought the problem was only for root tasks, as their stack is not allocated by Julia? Or what? Sorry, I am still confused about this :-/

Reply:
The Julia stack buffer information includes the guard pages. The information that Julia internally maintains is the start and end of the stack buffer, including guard pages. For some stacks, Julia may not even know where the guard pages end. This can in principle (with `COPY_STACKS`) be any task; in addition, it may change in future implementations.

I will try and spell out more clearly what we're doing here.